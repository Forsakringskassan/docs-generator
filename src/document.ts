import { type DocumentOutline } from "./document-outline";

/**
 * @internal
 */
export interface DocumentAttributes {
    title?: string;
    layout?: string;

    /** sets the `name` property in `Document` */
    name?: string;

    /** component status */
    status?: string;

    /** visible in navigation */
    visible?: boolean;

    /** if should write to file and make available as a page */
    include?: boolean;

    /** component(s) this page corresponds to */
    component?: string | Component | Array<string | Component>;

    /** link to external page in navigation */
    href?: string;

    /** navigation sortorder */
    sortorder?: number;
}

/**
 * @public
 */
export type DocumentBadge = "success" | "error" | "info";

export interface Component {
    /** component name */
    name: string;
    /** optional hint (glob pattern) how to find the component */
    source?: string;
}

/**
 * @public
 */
export interface NormalizedDocumentAttributes {
    title?: string;
    layout?: string;
    status?: string;
    badge?: DocumentBadge;
    component?: Component[];
    href?: string;
    /** normalized sortorder (defaults to Infinity) */
    sortorder: number;
}

/**
 * @public
 */
export interface FileInfo {
    /** path relative to configured base */
    path: string;
    /** filename without extension */
    name: string;
    /** path relative to project root */
    fullPath: string;
    /**
     * Output filename or `false` to disable writing result to a file.
     *
     * When passing a filename these optional placeholders can be used:
     *
     * - `[hash]` - replaced with the hash of the content body.
     */
    outputName: string | false;
}

/**
 * @public
 */
export interface Document {
    /** unique identifier for this document */
    id: string;

    /** human-readable pseudo-identifier for this document (consider it unique but there is no contraint enforcing this) */
    name: string;

    /** alternative names this document can be referenced by */
    alias: string[];

    /** true if document should be visible in menu and similar sources */
    visible: boolean;

    attributes: NormalizedDocumentAttributes;

    body: string;

    /** Document outline (i.e. the heading structure) */
    outline: DocumentOutline;

    /** format of body */
    format: "markdown" | "html" | "json";

    /*' list of generic tags */
    tags: string[];

    template: string;

    fileInfo: FileInfo;
}
