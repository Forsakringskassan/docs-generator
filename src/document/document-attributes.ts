/**
 * @public
 */
export type DocumentBadge = "success" | "error" | "info";

/**
 * @public
 */
export interface ComponentAttribute {
    /** component name */
    name: string;
    /** optional hint (glob pattern) how to find the component */
    source?: string;
}

/**
 * @internal
 */
export interface DocumentAttributes {
    /** Page title */
    title?: string;

    /** Short page title (e.g. used in sidenav) */
    "short-title"?: string;

    /** Page layout */
    layout?: string;

    /** sets the `name` property in `Document` */
    name?: string;

    /** document alias for easier linking */
    alias?: string | string[];

    /** component status */
    status?: string;

    /** visible in navigation */
    visible?: boolean;

    /** if should write to file and make available as a page */
    include?: boolean;

    /** component(s) this page corresponds to */
    component?:
        | string
        | ComponentAttribute
        | Array<string | ComponentAttribute>;

    /** link to external page in navigation */
    href?: string;

    /** navigation sortorder */
    sortorder?: number;

    /** pages redirecting to this page */
    redirect_from?: string | string[];
}

/**
 * @public
 */
export interface NormalizedDocumentAttributes {
    title?: string;
    shortTitle?: string;
    layout?: string;
    status?: string;
    badge?: DocumentBadge;
    component?: ComponentAttribute[];
    href?: string;
    /** normalized sortorder (defaults to Infinity) */
    sortorder: number;
    redirectFrom: string[];
}

/**
 * @internal
 */
export const documentAttributeKeys = [
    "title",
    "short-title",
    "layout",
    "name",
    "alias",
    "status",
    "visible",
    "include",
    "component",
    "href",
    "sortorder",
    "redirect_from",
];
