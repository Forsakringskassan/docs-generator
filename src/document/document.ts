import { type DocumentPage } from "./document-page";
import { type DocumentPartial } from "./document-partial";

/**
 * @public
 */
export type Document = DocumentPage | DocumentPartial;
