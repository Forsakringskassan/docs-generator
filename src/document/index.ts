export {
    type Component,
    type Document,
    type DocumentAttributes,
    type DocumentBadge,
    type FileInfo,
    type NormalizedDocumentAttributes,
    documentAttributeKeys,
} from "./document";
