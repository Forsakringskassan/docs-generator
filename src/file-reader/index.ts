export { type FileReader } from "./file-reader";
export { frontMatterFileReader } from "./front-matter-file-reader";
export { vueFileReader } from "./vue-file-reader";
export { type SourceFiles } from "./source-files";
export { fileReaderProcessor } from "./file-reader-processor";
export { navigationFileReader } from "./navigation-file-reader";
