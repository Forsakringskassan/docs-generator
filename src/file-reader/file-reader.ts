import { type Document } from "../document";

/**
 * Read and parse a file from disk to a doc-generator document.
 *
 * @public
 */
export type FileReader = (
    filePath: string,
    basePath?: string,
) => Promise<Document[]>;
