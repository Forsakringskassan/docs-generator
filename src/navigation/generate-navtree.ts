import path from "node:path";
import { type Document } from "../document";

type Mutable<T> = { -readonly [P in keyof T]: T[P] };

/**
 * @public
 */
export interface NavigationLeaf {
    readonly title: string;
    readonly path: string;
    readonly sortorder: number;
    readonly id: string;
    readonly external: boolean;
}

/**
 * @public
 */
export interface NavigationSection {
    readonly key: string;
    readonly title: string;
    readonly path: string;
    readonly sortorder: number;
    readonly children: NavigationNode[];
}

/**
 * @public
 */
export type NavigationNode = NavigationLeaf | NavigationSection;

type MutableNavigationSection = Mutable<NavigationSection>;

/**
 * @public
 */
export function isNavigationSection(
    node: NavigationNode,
): node is NavigationSection {
    return "key" in node;
}

function pathFromDoc({ fileInfo }: Document): [string, boolean] {
    if (fileInfo.name === "index") {
        return [fileInfo.path.replace(/\\/g, "/"), true];
    } else {
        return [
            `./${path.join(fileInfo.path, fileInfo.name).replace(/\\/g, "/")}`,
            false,
        ];
    }
}

function getParentKey(name: string): string {
    return name.split("/").slice(0, -1).join("/");
}

export function generateNavtree(docs: Document[]): NavigationSection {
    /** Lookup table for sections */
    const section: Record<string, MutableNavigationSection> = {};

    function createSection(
        key: string,
        title: string,
        sortorder: number,
    ): void {
        const existing = section[key];
        if (existing) {
            existing.title = title;
            existing.sortorder = sortorder;
        } else {
            const node: NavigationSection = {
                key,
                title,
                path: "",
                sortorder,
                children: [],
            };
            section[key] = node;
            if (key !== ".") {
                const parent = attach(key);
                parent.children.push(node);
            }
        }
    }

    function attach(key: string): MutableNavigationSection {
        const parentKey = getParentKey(key);
        let parent = section[parentKey];

        /* parent exists and is attached */
        if (parent) {
            return parent;
        }

        /* create stub if the parent does not exist (yet) */
        parent = {
            key: parentKey,
            title: parentKey.split("/").at(-1) ?? "(missing title)",
            path: "",
            sortorder: Infinity,
            children: [],
        };
        section[parentKey] = parent;
        let anchestor = parentKey;
        let current = parent;
        while (anchestor !== ".") {
            anchestor = getParentKey(anchestor);
            if (section[anchestor]) {
                section[anchestor].children.push(current);
                break;
            }
            const node: NavigationSection = {
                key: anchestor,
                title: "",
                path: "",
                sortorder: Infinity,
                children: [current],
            };
            section[anchestor] = node;
            current = node;
        }

        return parent;
    }

    for (const doc of docs) {
        const [name, isSection] = pathFromDoc(doc);
        const title =
            doc.attributes.shortTitle ??
            doc.attributes.title ??
            doc.fileInfo.name;
        const sortorder = doc.attributes.sortorder;

        /* external link */
        if (doc.attributes.href) {
            const parent = attach(name);
            const leafNode: NavigationLeaf = {
                id: doc.id,
                title,
                path: doc.attributes.href,
                sortorder,
                external: true,
            };
            parent.children.push(leafNode);
            continue;
        }

        /* ignore documents without any output */
        if (!doc.fileInfo.outputName) {
            if (isSection) {
                createSection(name, title, sortorder);
            }
            continue;
        }

        /* ignore documents with are marked as not navigatable */
        if (!doc.visible) {
            if (isSection) {
                createSection(name, title, sortorder);
            }
            continue;
        }

        /* special case: create or update root node */
        if (name === ".") {
            createSection(name, title, sortorder);
            continue;
        }

        const parent = attach(name);

        if (isSection) {
            const leafNode: NavigationLeaf = {
                id: doc.id,
                title,
                path: `${name}/`,
                sortorder,
                external: false,
            };

            const existing = section[name];
            if (existing) {
                existing.title = title;
                existing.path = leafNode.path;
                existing.sortorder = sortorder;
                existing.children.unshift(leafNode);
            } else {
                const sectionNode: NavigationSection = {
                    key: name,
                    title,
                    path: leafNode.path,
                    sortorder,
                    children: [leafNode],
                };
                section[name] = sectionNode;
                parent.children.push(sectionNode);
            }
        } else {
            const leafNode: NavigationLeaf = {
                id: doc.id,
                title,
                path: `${doc.fileInfo.path}/${doc.fileInfo.outputName}`,
                sortorder,
                external: false,
            };
            parent.children.push(leafNode);
        }
    }

    const root = section["."];
    if (root) {
        return root;
    } else {
        return {
            key: ".",
            title: "",
            path: "",
            sortorder: Infinity,
            children: [],
        };
    }
}
