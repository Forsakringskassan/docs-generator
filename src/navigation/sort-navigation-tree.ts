import { NavigationSection, isNavigationSection } from "./generate-navtree";

/**
 * @internal
 */
export function sortNavigationTree(tree: NavigationSection): void {
    tree.children.sort((a, b) => {
        if (a.sortorder < b.sortorder) {
            return -1;
        } else if (a.sortorder > b.sortorder) {
            return 1;
        }

        const titleA = a.title.toUpperCase();
        const titleB = b.title.toUpperCase();
        if (titleA < titleB) {
            return -1;
        }
        if (titleA > titleB) {
            return 1;
        }
        return 0;
    });
    for (const child of tree.children) {
        if (isNavigationSection(child)) {
            sortNavigationTree(child);
        }
    }
}
