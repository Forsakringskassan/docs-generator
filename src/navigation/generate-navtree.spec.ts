import path from "node:path";
import { glob } from "glob";
import { type Document } from "../document";
import { frontMatterFileReader } from "../file-reader";
import { generateNavtree } from "./generate-navtree";
import { sortNavigationTree } from "./sort-navigation-tree";

function createDocument(
    filePath: string,
    title: string,
    doc?: Partial<Document>,
): Document {
    const parsed = path.parse(filePath);
    return {
        id: `mock:${filePath}`,
        name: path.basename(filePath),
        alias: [],
        visible: true,
        body: "",
        outline: [],
        format: "markdown",
        tags: [],
        template: "mock",
        fileInfo: {
            path: parsed.dir !== "" ? `./${parsed.dir}` : ".",
            name: parsed.name,
            fullPath: filePath,
            outputName: `${parsed.name}.html`,
        },
        ...doc,
        attributes: {
            title,
            sortorder: Infinity /* sorting is tested separately */,
            ...doc?.attributes,
        },
    };
}

it("should create root node", () => {
    const nav = generateNavtree([]);
    expect(nav).toEqual({
        key: ".",
        title: "",
        path: "",
        sortorder: Infinity,
        visible: true,
        children: [],
    });
});

it("should merge root index.md to root node", () => {
    const nav = generateNavtree([createDocument("index.md", "Frontpage")]);
    expect(nav).toEqual({
        key: ".",
        title: "Frontpage",
        path: "",
        sortorder: Infinity,
        visible: true,
        children: [],
    });
});

it("should store sort order from document", () => {
    const nav = generateNavtree([
        createDocument("index.md", "Frontpage", {
            attributes: { sortorder: 10 },
        }),
    ]);
    expect(nav).toEqual(
        expect.objectContaining({
            title: "Frontpage",
            sortorder: 10,
        }),
    );
});

it("should use short title if present", () => {
    const nav = generateNavtree([
        createDocument("index.md", "Frontpage", {
            attributes: { shortTitle: "Short", sortorder: Infinity },
        }),
    ]);
    expect(nav).toEqual(
        expect.objectContaining({
            title: "Short",
        }),
    );
});

it("should ignore documents with visible false", () => {
    const nav = generateNavtree([
        createDocument("index.md", "Frontpage"),
        createDocument("foo.md", "Foo", { visible: false }),
    ]);
    expect(nav).toEqual({
        key: ".",
        title: "Frontpage",
        path: "",
        sortorder: Infinity,
        children: [],
        visible: true,
    });
});

it("should not ignore sections with visible false", () => {
    const nav = generateNavtree([
        createDocument("index.md", "Frontpage"),
        createDocument("foo/index.md", "Foo", { visible: false }),
        createDocument("foo/bar.md", "Bar"),
    ]);
    expect(nav).toEqual({
        key: ".",
        title: "Frontpage",
        path: "",
        sortorder: Infinity,
        visible: true,
        children: [
            {
                key: "./foo",
                title: "Foo",
                path: "",
                sortorder: Infinity,
                visible: false,
                children: [
                    {
                        id: "mock:foo/bar.md",
                        title: "Bar",
                        path: "./foo/bar.html",
                        sortorder: Infinity,
                        external: false,
                    },
                ],
            },
        ],
    });
});

it("should create children for each folder", () => {
    const nav = generateNavtree([
        createDocument("index.md", "Frontpage"),
        createDocument("usage/index.md", "Usage guide"),
        createDocument("usage/foo.md", "Foo component"),
        createDocument("components/index.md", "Components", {
            attributes: { sortorder: 10 },
        }),
    ]);
    expect(nav).toEqual({
        key: ".",
        title: "Frontpage",
        path: "",
        sortorder: Infinity,
        visible: true,
        children: [
            {
                key: "./usage",
                title: "Usage guide",
                path: "./usage/",
                sortorder: Infinity,
                visible: true,
                children: [
                    {
                        id: "mock:usage/index.md",
                        title: "Usage guide",
                        path: "./usage/",
                        sortorder: Infinity,
                        external: false,
                    },
                    {
                        id: "mock:usage/foo.md",
                        title: "Foo component",
                        path: "./usage/foo.html",
                        sortorder: Infinity,
                        external: false,
                    },
                ],
            },
            {
                key: "./components",
                title: "Components",
                path: "./components/",
                sortorder: 10,
                visible: true,
                children: [
                    {
                        id: "mock:components/index.md",
                        path: "./components/",
                        title: "Components",
                        sortorder: 10,
                        external: false,
                    },
                ],
            },
        ],
    });
});

it("should create children for each folder (inverse order)", () => {
    const nav = generateNavtree([
        createDocument("components/index.md", "Components"),
        createDocument("usage/foo.md", "Foo component"),
        createDocument("usage/index.md", "Usage guide", {
            attributes: { sortorder: 20 },
        }),
        createDocument("index.md", "Frontpage", {
            attributes: { sortorder: 10 },
        }),
    ]);
    expect(nav).toEqual({
        key: ".",
        title: "Frontpage",
        path: "",
        sortorder: 10,
        visible: true,
        children: [
            {
                key: "./components",
                title: "Components",
                path: "./components/",
                sortorder: Infinity,
                visible: true,
                children: [
                    {
                        id: "mock:components/index.md",
                        path: "./components/",
                        title: "Components",
                        sortorder: Infinity,
                        external: false,
                    },
                ],
            },
            {
                key: "./usage",
                title: "Usage guide",
                path: "./usage/",
                sortorder: 20,
                visible: true,
                children: [
                    {
                        id: "mock:usage/index.md",
                        title: "Usage guide",
                        path: "./usage/",
                        sortorder: 20,
                        external: false,
                    },
                    {
                        id: "mock:usage/foo.md",
                        title: "Foo component",
                        path: "./usage/foo.html",
                        sortorder: Infinity,
                        external: false,
                    },
                ],
            },
        ],
    });
});

it("should create children for subpages", () => {
    const nav = generateNavtree([
        createDocument("index.md", "Frontpage"),
        createDocument("a/index.md", "A"),
        createDocument("a/foo.md", "A > Foo"),
        createDocument("a/bar.md", "A > Bar"),
        createDocument("b/index.md", "B"),
        createDocument("b/foo.md", "B > Foo"),
        createDocument("b/bar.md", "B > Bar"),
        createDocument("b/c/index.md", "B > C"),
        createDocument("b/c/foo.md", "B > C > Foo"),
        createDocument("b/c/bar.md", "B > C > Bar"),
    ]);
    expect(nav).toEqual({
        key: ".",
        title: "Frontpage",
        path: "",
        sortorder: Infinity,
        visible: true,
        children: [
            {
                key: "./a",
                title: "A",
                path: "./a/",
                sortorder: Infinity,
                visible: true,
                children: [
                    {
                        id: "mock:a/index.md",
                        path: "./a/",
                        sortorder: Infinity,
                        title: "A",
                        external: false,
                    },
                    {
                        id: "mock:a/foo.md",
                        path: "./a/foo.html",
                        sortorder: Infinity,
                        title: "A > Foo",
                        external: false,
                    },
                    {
                        id: "mock:a/bar.md",
                        path: "./a/bar.html",
                        sortorder: Infinity,
                        title: "A > Bar",
                        external: false,
                    },
                ],
            },
            {
                key: "./b",
                title: "B",
                path: "./b/",
                sortorder: Infinity,
                visible: true,
                children: [
                    {
                        id: "mock:b/index.md",
                        path: "./b/",
                        sortorder: Infinity,
                        title: "B",
                        external: false,
                    },
                    {
                        id: "mock:b/foo.md",
                        path: "./b/foo.html",
                        sortorder: Infinity,
                        title: "B > Foo",
                        external: false,
                    },
                    {
                        id: "mock:b/bar.md",
                        path: "./b/bar.html",
                        sortorder: Infinity,
                        title: "B > Bar",
                        external: false,
                    },
                    {
                        key: "./b/c",
                        title: "B > C",
                        path: "./b/c/",
                        sortorder: Infinity,
                        visible: true,
                        children: [
                            {
                                id: "mock:b/c/index.md",
                                path: "./b/c/",
                                sortorder: Infinity,
                                title: "B > C",
                                external: false,
                            },
                            {
                                id: "mock:b/c/foo.md",
                                path: "./b/c/foo.html",
                                sortorder: Infinity,
                                title: "B > C > Foo",
                                external: false,
                            },
                            {
                                id: "mock:b/c/bar.md",
                                path: "./b/c/bar.html",
                                sortorder: Infinity,
                                title: "B > C > Bar",
                                external: false,
                            },
                        ],
                    },
                ],
            },
        ],
    });
});

it("should read title and sortorder from hidden index.md", () => {
    const nav = generateNavtree([
        createDocument("usage/index.md", "Usage guide", {
            visible: false,
            attributes: { sortorder: 10 },
        }),
        createDocument("usage/foo.md", "Foo component"),
    ]);
    expect(nav).toEqual({
        key: ".",
        title: ".",
        path: "",
        sortorder: Infinity,
        visible: true,
        children: [
            {
                key: "./usage",
                title: "Usage guide",
                path: "",
                sortorder: 10,
                visible: false,
                children: [
                    {
                        id: "mock:usage/foo.md",
                        title: "Foo component",
                        path: "./usage/foo.html",
                        sortorder: Infinity,
                        external: false,
                    },
                ],
            },
        ],
    });
});

async function getDocsFromGlob(
    pattern: string,
    basePath: string,
): Promise<Document[]> {
    const filenames = await glob(pattern, {
        ignore: "./docs/node_modules/**",
        nodir: true,
    });
    const parsed = filenames.map((it) => frontMatterFileReader(it, basePath));
    return (await Promise.all(parsed)).flat();
}

it("smoketest", async () => {
    const docs = await getDocsFromGlob("./docs/**/*.md", "./docs");
    const nav = generateNavtree(docs);
    sortNavigationTree(nav); // to prevent flakyness
    expect(nav).toMatchSnapshot();
});
