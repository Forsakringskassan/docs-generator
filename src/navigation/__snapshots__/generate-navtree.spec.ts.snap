// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoketest 1`] = `
{
  "children": [
    {
      "children": [
        {
          "external": false,
          "id": "fs:docs/markdown/index.md",
          "path": "./markdown/",
          "sortorder": 0,
          "title": "Markdown",
        },
        {
          "children": [
            {
              "external": false,
              "id": "fs:docs/markdown/code-preview/code-preview.md",
              "path": "./markdown/code-preview/",
              "sortorder": Infinity,
              "title": "Code preview",
            },
            {
              "external": false,
              "id": "fs:docs/markdown/code-preview/fullscreen.md",
              "path": "./markdown/code-preview/fullscreen.html",
              "sortorder": Infinity,
              "title": "Fullscreen examples",
            },
          ],
          "key": "./markdown/code-preview",
          "path": "./markdown/code-preview/",
          "sortorder": Infinity,
          "title": "Code preview",
        },
        {
          "external": false,
          "id": "fs:docs/markdown/custom-containers.md",
          "path": "./markdown/custom-containers.html",
          "sortorder": Infinity,
          "title": "Custom containers",
        },
        {
          "external": false,
          "id": "fs:docs/markdown/mermaid.md",
          "path": "./markdown/mermaid.html",
          "sortorder": Infinity,
          "title": "Mermaid",
        },
      ],
      "key": "./markdown",
      "path": "./markdown/",
      "sortorder": 0,
      "title": "Markdown",
    },
    {
      "children": [
        {
          "external": false,
          "id": "fs:docs/navigation/index.md",
          "path": "./navigation/",
          "sortorder": 1,
          "title": "Navigation",
        },
        {
          "external": false,
          "id": "fs:docs/navigation/child-page-1.md",
          "path": "./navigation/child-page-1.html",
          "sortorder": Infinity,
          "title": "Child Page 1",
        },
        {
          "external": false,
          "id": "fs:docs/navigation/child-page-2.md",
          "path": "./navigation/child-page-2.html",
          "sortorder": Infinity,
          "title": "Child Page 2",
        },
        {
          "external": false,
          "id": "fs:docs/navigation/child-page-3.md",
          "path": "./navigation/child-page-3.html",
          "sortorder": Infinity,
          "title": "Child Page 3",
        },
        {
          "children": [
            {
              "external": false,
              "id": "fs:docs/navigation/group/nested-page-1.md",
              "path": "./navigation/group/nested-page-1.html",
              "sortorder": Infinity,
              "title": "Nested Page 1",
            },
            {
              "external": false,
              "id": "fs:docs/navigation/group/nested-page-2.md",
              "path": "./navigation/group/nested-page-2.html",
              "sortorder": Infinity,
              "title": "Nested Page 2",
            },
            {
              "external": false,
              "id": "fs:docs/navigation/group/nested-page-3.md",
              "path": "./navigation/group/nested-page-3.html",
              "sortorder": Infinity,
              "title": "Nested Page 3",
            },
          ],
          "key": "./navigation/group",
          "path": "",
          "sortorder": Infinity,
          "title": "group",
        },
        {
          "children": [
            {
              "external": false,
              "id": "fs:docs/navigation/sorted-group/top-page.md",
              "path": "./navigation/sorted-group/top-page.html",
              "sortorder": 1,
              "title": "Top page",
            },
            {
              "external": false,
              "id": "fs:docs/navigation/sorted-group/sorted-page-3.md",
              "path": "./navigation/sorted-group/sorted-page-3.html",
              "sortorder": 2,
              "title": "Sorted Page 3",
            },
            {
              "external": false,
              "id": "fs:docs/navigation/sorted-group/sorted-page-1.md",
              "path": "./navigation/sorted-group/sorted-page-1.html",
              "sortorder": Infinity,
              "title": "Sorted Page 1",
            },
            {
              "external": false,
              "id": "fs:docs/navigation/sorted-group/sorted-page-2.md",
              "path": "./navigation/sorted-group/sorted-page-2.html",
              "sortorder": Infinity,
              "title": "Sorted Page 2",
            },
          ],
          "key": "./navigation/sorted-group",
          "path": "",
          "sortorder": Infinity,
          "title": "sorted-group",
        },
      ],
      "key": "./navigation",
      "path": "./navigation/",
      "sortorder": 1,
      "title": "Navigation",
    },
    {
      "children": [
        {
          "external": false,
          "id": "fs:docs/processors/index.md",
          "path": "./processors/",
          "sortorder": 1,
          "title": "Processors",
        },
        {
          "external": false,
          "id": "fs:docs/processors/cookie.md",
          "path": "./processors/cookie.html",
          "sortorder": Infinity,
          "title": "Cookie",
        },
        {
          "external": false,
          "id": "fs:docs/processors/manifest.md",
          "path": "./processors/manifest.html",
          "sortorder": Infinity,
          "title": "Manifest",
        },
        {
          "external": false,
          "id": "fs:docs/processors/motd.md",
          "path": "./processors/motd.html",
          "sortorder": Infinity,
          "title": "MOTD",
        },
        {
          "external": false,
          "id": "fs:docs/processors/nunjucks.md",
          "path": "./processors/nunjucks.html",
          "sortorder": Infinity,
          "title": "Nunjucks",
        },
        {
          "external": false,
          "id": "fs:docs/processors/source-url.md",
          "path": "./processors/source-url.html",
          "sortorder": Infinity,
          "title": "Source url",
        },
        {
          "external": false,
          "id": "fs:docs/processors/topnav.md",
          "path": "./processors/topnav.html",
          "sortorder": Infinity,
          "title": "Topnav",
        },
        {
          "external": false,
          "id": "fs:docs/processors/version.md",
          "path": "./processors/version.html",
          "sortorder": Infinity,
          "title": "Version",
        },
      ],
      "key": "./processors",
      "path": "./processors/",
      "sortorder": 1,
      "title": "Processors",
    },
    {
      "children": [
        {
          "external": false,
          "id": "fs:docs/theme/index.md",
          "path": "./theme/",
          "sortorder": 1,
          "title": "Theme",
        },
      ],
      "key": "./theme",
      "path": "./theme/",
      "sortorder": 1,
      "title": "Theme",
    },
    {
      "children": [
        {
          "external": false,
          "id": "fs:docs/live-example/example.md",
          "path": "./live-example/example.html",
          "sortorder": Infinity,
          "title": "Example",
        },
      ],
      "key": "./live-example",
      "path": "",
      "sortorder": Infinity,
      "title": "live-example",
    },
  ],
  "key": ".",
  "path": "",
  "sortorder": Infinity,
  "title": "Todo",
}
`;
